README.md
go-httpdate
=======

[![Build Status](https://travis-ci.org/Songmu/go-httpdate.png?branch=master)][travis]
[![Coverage Status](https://coveralls.io/repos/Songmu/go-httpdate/badge.png?branch=master)][coveralls]
[![MIT License](http://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)][license]
[![GoDoc](https://godoc.org/github.com/Songmu/go-httpdate?status.svg)][godoc]

[travis]: https://travis-ci.org/Songmu/go-httpdate
[coveralls]: https://coveralls.io/r/Songmu/go-httpdate?branch=master
[license]: https://github.com/Songmu/go-httpdate/blob/master/LICENSE
[godoc]: https://godoc.org/github.com/Songmu/go-httpdate

## Description

Well dealing the date formats used by the HTTP protocol (and then some more).  
`Str2Time` is tremendous function which can detect various date formats from string and returns `time.Time`.

Is is golang porting of perl's [HTTP::Date](https://metacpan.org/pod/HTTP::Date) (ported only `str2time` interface)

## Synopsis

    # try to parse string and returns `time.Time` or error
    t1, err := httpdate.Str2Time("Thu, 03 Feb 1994 12:33:44 GMT", time.UTC)
    t2, err := httpdate.Str2Time("2017-11-11", time.UTC)
    t3, err := httpdate.Str2Time("Thu Nov  9 18:20:31 GMT 2017", time.UTC)
    t4, err := httpdate.Str2Time("08-Feb-94 14:15:29 GMT", time.UTC)

## Supported Format

-  "Wed, 09 Feb 1994 22:23:32 GMT"       -- HTTP format
-  "Thu Feb  3 17:03:55 GMT 1994"        -- ctime(3) format
-  "Thu Feb  3 00:00:00 1994",           -- ANSI C asctime() format
-  "Tuesday, 08-Feb-94 14:15:29 GMT"     -- old rfc850 HTTP format
-  "Tuesday, 08-Feb-1994 14:15:29 GMT"   -- broken rfc850 HTTP format
-  "03/Feb/1994:17:03:55 -0700"   -- common logfile format
-  "09 Feb 1994 22:23:32 GMT"     -- HTTP format (no weekday)
-  "08-Feb-94 14:15:29 GMT"       -- rfc850 format (no weekday)
-  "08-Feb-1994 14:15:29 GMT"     -- broken rfc850 format (no weekday)
-  "1994-02-03 14:15:29 -0100"    -- ISO 8601 format
-  "1994-02-03 14:15:29"          -- zone is optional
-  "1994-02-03"                   -- only date
-  "1994-02-03T14:15:29"          -- Use T as separator
-  "19940203T141529Z"             -- ISO 8601 compact format
-  "19940203"                     -- only date
-  "08-Feb-94"         -- old rfc850 HTTP format    (no weekday, no time)
-  "08-Feb-1994"       -- broken rfc850 HTTP format (no weekday, no time)
-  "09 Feb 1994"       -- proposed new HTTP format  (no weekday, no time)
-  "03/Feb/1994"       -- common logfile format     (no time, no offset)
-  "Feb  3  1994"      -- Unix 'ls -l' format
-  "Feb  3 17:03"      -- Unix 'ls -l' format
-  "11-15-96  03:52PM" -- Windows 'dir' format

## See Also

- [HTTP::Date](https://metacpan.org/pod/HTTP::Date)

## Author

[Songmu](https://github.com/Songmu)


CHANGELOG.md
# Changelog

## [v1.0.0](https://github.com/Songmu/go-httpdate/compare/v0.0.1...v1.0.0) (2017-11-13)

* ready for v1 release [#2](https://github.com/Songmu/go-httpdate/pull/2) ([Songmu](https://github.com/Songmu))

## [v0.0.1](https://github.com/Songmu/go-httpdate/compare/ff508c6...v0.0.1) (2017-11-13)

* Initial release [#1](https://github.com/Songmu/go-httpdate/pull/1) ([Songmu](https://github.com/Songmu))


example_test.go
package httpdate_test

import (
	"fmt"
	"time"

	httpdate "github.com/Songmu/go-httpdate"
)

func ExampleStr2Time() {
	t1, _ := httpdate.Str2Time("Thu, 03 Feb 1994 12:33:44 GMT", time.UTC)
	t2, _ := httpdate.Str2Time("2017-11-11", time.UTC)
	t3, _ := httpdate.Str2Time("Thu Nov  9 18:20:31 GMT 2017", time.UTC)
	t4, _ := httpdate.Str2Time("08-Feb-94 14:15:29 GMT", time.UTC)

	fmt.Println(t1)
	fmt.Println(t2)
	fmt.Println(t3)
	fmt.Println(t4)
	// Output:
	// 1994-02-03 12:33:44 +0000 GMT
	// 2017-11-11 00:00:00 +0000 UTC
	// 2017-11-09 18:20:31 +0000 GMT
	// 1994-02-08 14:15:29 +0000 GMT
}


httpdate_test.go
package httpdate

import (
	"fmt"
	"reflect"
	"testing"
	"time"
)

func TestStr2Time(t *testing.T) {
	expect := time.Date(1994, time.February, 3, 0, 0, 0, 0, time.UTC)
	testCases := []struct {
		name  string
		input string
	}{
		{
			name:  "RFC1123",
			input: "Thu, 03 Feb 1994 00:00:00 GMT",
		},
		{
			name:  "old rfc850 HTTP format",
			input: "Thursday, 03-Feb-94 00:00:00 GMT",
		},
		{
			name:  "broken rfc850 HTTP format",
			input: "Thursday, 03-Feb-1994 00:00:00 GMT",
		},
		{
			name:  "common logfile format1",
			input: "03/Feb/1994:00:00:00 0000",
		},
		{
			name:  "common logfile format2",
			input: "03/Feb/1994:01:00:00 +0100",
		},
		{
			name:  "common logfile format1",
			input: "02/Feb/1994:23:00:00 -0100",
		},
		{
			name:  "HTTP format (no weekday)",
			input: "03 Feb 1994 00:00:00 GMT",
		},
		{
			name:  "old rfc850 (no weekday)",
			input: "03-Feb-94 00:00:00 GMT",
		},
		{
			name:  "broken rfc850 (no weekday)",
			input: "03-Feb-1994 00:00:00 GMT",
		},
		{
			name:  "broken rfc850 (no weekday, no seconds)",
			input: "03-Feb-1994 00:00 GMT",
		},
		{
			name:  "VMS dir listing format",
			input: "03-Feb-1994 00:00",
		},
		{
			name:  "old rfc850 HTTP format (no weekday, no time)",
			input: "03-Feb-94",
		},
		{
			name:  "broken rfc850 HTTP format (no weekday, no time)",
			input: "03-Feb-1994",
		},
		{
			name:  "proposed new HTTP format (no weekday, no time)",
			input: "03 Feb 1994",
		},
		{
			name:  "common logfile format (no time, no offset)",
			input: "03/Feb/1994",
		},
		{
			name:  "A few tests with extra space at various places 1",
			input: "  03/Feb/1994      ",
		},
		{
			name:  "A few tests with extra space at various places 2",
			input: "  03   Feb   1994  0:00  ",
		},
		{
			name:  "Tests a commonly used (faulty?) date format of php cms systems",
			input: "Thu, 03 Feb 1994 00:00:00 +0000 GMT",
		},
		{
			name:  "ctime format",
			input: "Thu Feb  3 00:00:00 GMT 1994",
		},
		{
			name:  "same as ctime, except no TZ",
			input: "Thu Feb  3 00:00:00 1994",
		},
		{
			name:  "Unix 'ls -l' format",
			input: "Feb  3 1994",
		},
		{
			name:  "ISO 8601 formats 1",
			input: "1994-02-03 00:00:00 +0000",
		},
		{
			name:  "ISO 8601 formats 2",
			input: "1994-02-03",
		},
		{
			name:  "ISO 8601 formats 3",
			input: "19940203",
		},
		{
			name:  "ISO 8601 formats 4",
			input: "1994-02-03T00:00:00+0000",
		},
		{
			name:  "ISO 8601 formats 5",
			input: "1994-02-02T23:00:00-0100",
		},
		{
			name:  "ISO 8601 formats 6",
			input: "1994-02-02T23:00:00-01:00",
		},
		{
			name:  "ISO 8601 formats 7",
			input: "1994-02-03T00:00:00 Z",
		},
		{
			name:  "ISO 8601 formats 8",
			input: "19940203T000000Z",
		},
		{
			name:  "ISO 8601 formats 9",
			input: "199402030000",
		},
		{
			name:  "Windows 'dir' format",
			input: "02-03-94  12:00AM",
		},
	}
	for _, tc := range testCases {
		out, err := Str2Time(tc.input, time.UTC)
		if err != nil {
			t.Errorf("%s error should be nil but: %s", tc.name, err)
		}
		out = out.In(time.UTC)
		if !reflect.DeepEqual(out, expect) {
			t.Errorf("Parse failed(%s):\n out:  %+v\n want: %+v", tc.name, out, expect)
		}
	}
}

func TestStr2Time_time(t *testing.T) {
	expect := time.Date(1994, time.February, 3, 4, 56, 1, 0, time.UTC)
	testCases := []struct {
		name  string
		input string
	}{
		{
			name:  "RFC1123",
			input: "Thu, 03 Feb 1994 04:56:01 GMT",
		},
		{
			name:  "old rfc850 HTTP format",
			input: "Thursday, 03-Feb-94 04:56:01 GMT",
		},
		{
			name:  "broken rfc850 HTTP format",
			input: "Thursday, 03-Feb-1994 04:56:01 GMT",
		},
		{
			name:  "common logfile format1",
			input: "03/Feb/1994:04:56:01 0000",
		},
		{
			name:  "common logfile format2",
			input: "03/Feb/1994:05:56:01 +0100",
		},
		{
			name:  "common logfile format1",
			input: "03/Feb/1994:03:56:01 -0100",
		},
		{
			name:  "HTTP format (no weekday)",
			input: "03 Feb 1994 04:56:01 GMT",
		},
		{
			name:  "old rfc850 (no weekday)",
			input: "03-Feb-94 04:56:01 GMT",
		},
		{
			name:  "broken rfc850 (no weekday)",
			input: "03-Feb-1994 04:56:01 GMT",
		},
		{
			name:  "Tests a commonly used (faulty?) date format of php cms systems",
			input: "Thu, 03 Feb 1994 04:56:01 +0000 GMT",
		},
		{
			name:  "ctime format",
			input: "Thu Feb  3 04:56:01 GMT 1994",
		},
		{
			name:  "same as ctime, except no TZ",
			input: "Thu Feb  3 04:56:01 1994",
		},
		{
			name:  "ISO 8601 formats 1",
			input: "1994-02-03 04:56:01 +0000",
		},
		{
			name:  "ISO 8601 formats 4",
			input: "1994-02-03T04:56:01+0000",
		},
		{
			name:  "ISO 8601 formats 5",
			input: "1994-02-03T03:56:01-0100",
		},
		{
			name:  "ISO 8601 formats 6",
			input: "1994-02-03T03:56:01-01:00",
		},
		{
			name:  "ISO 8601 formats 7",
			input: "1994-02-03T04:56:01 Z",
		},
		{
			name:  "ISO 8601 formats 8",
			input: "19940203T045601Z",
		},
	}
	for _, tc := range testCases {
		out, err := Str2Time(tc.input, time.UTC)
		if err != nil {
			t.Errorf("%s error should be nil but: %s", tc.name, err)
		}
		out = out.In(time.UTC)
		if !reflect.DeepEqual(out, expect) {
			t.Errorf("Parse failed(%s):\n out:  %+v\n want: %+v", tc.name, out, expect)
		}
	}
}

func TestStr2Time_noSec(t *testing.T) {
	y := time.Now().Year()
	expect := time.Date(y, time.February, 3, 16, 56, 0, 0, time.UTC)
	testCases := []struct {
		name  string
		input string
	}{
		{
			name:  "broken rfc850 (no weekday, no seconds)",
			input: fmt.Sprintf("03-Feb-%d 16:56 GMT", y),
		},
		{
			name:  "VMS dir listing format",
			input: fmt.Sprintf("03-Feb-%d 16:56", y),
		},
		{
			name:  "A few tests with extra space at various places 2",
			input: fmt.Sprintf("  03   Feb   %d  16:56  ", y),
		},
		{
			name:  "Unix 'ls -l' format",
			input: "Feb  3 16:56",
		},
		{
			name:  "ISO 8601 formats 9",
			input: fmt.Sprintf("%d02031656", y),
		},
		{
			name:  "Windows 'dir' format",
			input: fmt.Sprintf("02-03-%d   4:56PM", y-2000),
		},
	}
	for _, tc := range testCases {
		out, err := Str2Time(tc.input, time.UTC)
		if err != nil {
			t.Errorf("%s error should be nil but: %s", tc.name, err)
		}
		out = out.In(time.UTC)
		if !reflect.DeepEqual(out, expect) {
			t.Errorf("Parse failed(%s/%s):\n out:  %+v\n want: %+v", tc.name, tc.input, out, expect)
		}
	}
}

func TestStr2Time_nsec(t *testing.T) {
	testCases := []struct {
		name  string
		input string
		want  time.Time
	}{
		{
			name:  "Nine 9 nsec",
			input: "2006-01-02T15:04:05.999999999+00:00",
			want:  time.Date(2006, 1, 2, 15, 4, 5, 999999999, time.UTC),
		},
		{
			name:  "Third place after the decimal point",
			input: "2006-01-02T15:04:05.012Z",
			want:  time.Date(2006, 1, 2, 15, 4, 5, 12000000, time.UTC),
		},
	}
	for _, tc := range testCases {
		out, err := Str2Time(tc.input, time.UTC)
		if err != nil {
			t.Errorf("%s error should be nil but: %s", tc.name, err)
		}
		out = out.In(time.UTC)
		if !reflect.DeepEqual(out, tc.want) {
			t.Errorf("Parse failed(%s/%s):\n out:  %+v\n want: %+v", tc.name, tc.input, out, tc.want)
		}
	}
}


cli.go
package httpdate

import (
	"flag"
	"fmt"
	"io"
	"os"
	"regexp"
	"time"
)

const (
	exitcodeOK = iota
	exitCodeParseFlagErr
	exitCodeErr
)

// Run the cli
func Run(argv []string) int {
	return (&cli{os.Stdout, os.Stderr}).run(argv)
}

type cli struct {
	outStream, errStream io.Writer
}

var numOnlyReg = regexp.MustCompile(`^\d+$`)

func (cl *cli) run(argv []string) int {
	fs := flag.NewFlagSet("httpdate", flag.ContinueOnError)
	fs.SetOutput(cl.errStream)
	fs.Usage = func() {
		fmt.Fprintf(cl.errStream, "Usage of httpdate version %s (rev: %s):\n", Version, revision)
		fs.PrintDefaults()
	}
	var str2time = fs.Bool("s", false, "force str2time")
	if err := fs.Parse(argv); err != nil {
		if err == flag.ErrHelp {
			return exitcodeOK
		}
		return exitCodeParseFlagErr
	}
	if fs.NArg() != 1 {
		fs.Usage()
		return exitCodeParseFlagErr
	}

	arg := fs.Args()[0]
	if !*str2time && !numOnlyReg.MatchString(arg) {
		*str2time = true
	}

	if *str2time {
		t, err := Str2Time(arg, nil)
		if err != nil {
			fmt.Fprintln(cl.errStream, err)
			return exitCodeErr
		}
		fmt.Fprintf(cl.outStream, "%d\n", t.Unix())
	} else {
		s := Time2Str(time.Unix(int64(a2i(arg)), 0))
		fmt.Fprintln(cl.outStream, s)
	}
	return exitcodeOK
}


httpdate.go
package httpdate

import (
	"fmt"
	"math"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var (
	// Thu, 03 Feb 1994 00:00:00 GMT
	fastReg        = regexp.MustCompile(`^[SMTWF][a-z][a-z], (\d\d) ([JFMAJSOND][a-z][a-z]) (\d\d\d\d) (\d\d):(\d\d):(\d\d) GMT$`)
	uselessWdayReg = regexp.MustCompile(`^(?i)(?:Sun|Mon|Tue|Wed|Thu|Fri|Sat)[a-z]*,?\s*`)
	mostFormatReg  = regexp.MustCompile(`^(\d\d?)` + // 1. day
		`(?:\s+|[-\/])` +
		`(\w+)` + // 2. month
		`(?:\s+|[-\/])` +
		`(\d+)` + // 3. year
		`(?:` +
		`(?:\s+|:)` + // separator before clock
		`(\d\d?):(\d\d)` + // 4. 5. hour:min
		`(?::(\d\d))?` + // 6. optional seconds
		`)?\s*` + // optional clock
		`([-+]?\d{2,4})?\s*` + // 7. offset
		`(\w+)?` + // 8. ASCII representation of timezone.
		`\s*$`)
	ctimeAndAsctimeReg = regexp.MustCompile(`^(\w{1,3})\s+` + // 1. month
		`(\d\d?)\s+` + // 2. day
		`(\d\d?):(\d\d)` + // 3,4. hour:min
		`(?::(\d\d))?\s+` + // 5. optional seconds
		`(?:([A-Za-z]+)\s+)?` + // 6. optional timezone
		`(\d+)` + // 7. year
		`\s*$`)
	unixLsReg = regexp.MustCompile(`^(\w{3})\s+` + // 1. month
		`(\d\d?)\s+` + // 2. day
		`(?:` +
		`(\d{4})|` + // 3. year(optional)
		`(\d{1,2}):(\d{2})` + // 4,5. hour:min
		`(?::(\d{2}))?` + // 6 optional seconds
		`)\s*$`)
	iso8601Reg = regexp.MustCompile(`^(\d{4})` + // 1. year
		`[-\/]?` +
		`(\d\d?)` + // 2. numerical month
		`[-\/]?` +
		`(\d\d?)` + // 3. day
		`(?:` +
		`(?:\s*|[-:Tt])` + // separator before clock
		`(\d\d?):?(\d\d)` + // 4,5. hour:min
		`(?::?(\d\d)(?:\.(\d+))?)?` + // 6,7. optional seconds and fractional
		`)?\s*` + // optional clock
		`([-+]?\d\d?:?(:?\d\d)?|Z|z)?` + // 8. offset (Z is "zero meridian", i.e. GMT)
		`\s*$`)
	winDirReg = regexp.MustCompile(`^(\d{2})-` + // 1. mumerical month
		`(\d{2})-` + // 2. day
		`(\d{2})\s+` + // 3. year
		`(\d\d?):(\d\d)([APap][Mm])` + // 4,5,6. hour:min AM/PM
		`\s*$`)
	looksLikeTZReg = regexp.MustCompile("^[A-Z][A-Za-z]{2}")
)

var shortMon2Mon = map[string]time.Month{
	"Jan": time.January,
	"Feb": time.February,
	"Mar": time.March,
	"Apr": time.April,
	"May": time.May,
	"Jun": time.June,
	"Jul": time.July,
	"Aug": time.August,
	"Sep": time.September,
	"Oct": time.October,
	"Nov": time.November,
	"Dec": time.December,
}

func a2i(str string) int {
	i, _ := strconv.Atoi(str)
	return i
}

var offsetStrReg = regexp.MustCompile(`^([-+])?(\d\d?):?(\d\d)?$`)

func offsetStr2offset(str string) int {
	if m := offsetStrReg.FindStringSubmatch(str); len(m) > 0 {
		hour := a2i(m[2])
		min := a2i(m[3])
		offset := hour*60*60 + min*60
		if m[1] == "-" {
			offset *= -1
		}
		return offset
	}
	return 0
}

// Str2Time will try to convert a date string, and then return it as a
// `time.Time`. If the date is unrecognized, error will be returned.
//
// The function is able to parse the following formats:
//
//  "Wed, 09 Feb 1994 22:23:32 GMT"       -- HTTP format
//  "Thu Feb  3 17:03:55 GMT 1994"        -- ctime(3) format
//  "Thu Feb  3 00:00:00 1994",           -- ANSI C asctime() format
//  "Tuesday, 08-Feb-94 14:15:29 GMT"     -- old rfc850 HTTP format
//  "Tuesday, 08-Feb-1994 14:15:29 GMT"   -- broken rfc850 HTTP format
//
//  "03/Feb/1994:17:03:55 -0700"   -- common logfile format
//  "09 Feb 1994 22:23:32 GMT"     -- HTTP format (no weekday)
//  "08-Feb-94 14:15:29 GMT"       -- rfc850 format (no weekday)
//  "08-Feb-1994 14:15:29 GMT"     -- broken rfc850 format (no weekday)
//
//  "1994-02-03 14:15:29 -0100"    -- ISO 8601 format
//  "1994-02-03 14:15:29"          -- zone is optional
//  "1994-02-03"                   -- only date
//  "1994-02-03T14:15:29"          -- Use T as separator
//  "19940203T141529Z"             -- ISO 8601 compact format
//  "19940203"                     -- only date
//
//  "08-Feb-94"         -- old rfc850 HTTP format    (no weekday, no time)
//  "08-Feb-1994"       -- broken rfc850 HTTP format (no weekday, no time)
//  "09 Feb 1994"       -- proposed new HTTP format  (no weekday, no time)
//  "03/Feb/1994"       -- common logfile format     (no time, no offset)
//
//  "Feb  3  1994"      -- Unix 'ls -l' format
//  "Feb  3 17:03"      -- Unix 'ls -l' format
//
//  "11-15-96  03:52PM" -- Windows 'dir' format
//
// The parser ignores leading and trailing whitespace.  It also allow the
// seconds to be missing and the month to be numerical in most formats.
//
// If the year is missing, then we assume that the date is the first
// matching date I<before> current month.  If the year is given with only
// 2 digits, then parse_date() will select the century that makes the
// year closest to the current date.
//
// If no timezones are detected from string, location specified by 2nd argument
// is used. When the location is also nil, time.Local used.
// Please note that the returned `time.Time` does not always have the location
// specified. It is used only when the timezone can not be detected.
// If you want to fix the location in the returned time, please specify
// the location by yourself as follows.
//
//     t = t.In(time.Local)
func Str2Time(origTimeStr string, loc *time.Location) (time.Time, error) {
	if loc == nil {
		loc = time.Local
	}
	if m := fastReg.FindStringSubmatch(origTimeStr); len(m) > 0 {
		return time.Date(a2i(m[3]), shortMon2Mon[m[2]], a2i(m[1]), a2i(m[4]), a2i(m[5]), a2i(m[6]), 0, time.FixedZone("GMT", 0)), nil
	}
	timeStr := strings.TrimSpace(origTimeStr)
	timeStr = uselessWdayReg.ReplaceAllString(timeStr, "")

	adjustYear := func(str string) int {
		y := a2i(str)
		switch {
		case y >= 100:
			return y
		case y >= 69: // Unix time starts Dec 31 1969 in some time zones
			return y + 1900
		}
		return y + 2000
	}

	loadLoc := func(str string) *time.Location {
		switch strings.ToUpper(str) {
		case "Z":
			return time.UTC
		case "UT":
			str = "UTC"
		}
		l, _ := time.LoadLocation(str)
		return l
	}

	if m := mostFormatReg.FindStringSubmatch(timeStr); len(m) > 0 {
		switch strings.ToUpper(m[8]) {
		case "AM", "PM":
			// nop and through the next check
		default:
			var l *time.Location
			if m[8] != "" {
				l = loadLoc(m[8])
			}
			if l == nil && m[7] != "" {
				locName := m[8]
				if !looksLikeTZReg.MatchString(locName) {
					locName = ""
				}
				l = time.FixedZone(locName, offsetStr2offset(m[7]))
			}
			if l == nil {
				l = loc
			}
			return time.Date(adjustYear(m[3]), shortMon2Mon[m[2]], a2i(m[1]), a2i(m[4]), a2i(m[5]), a2i(m[6]), 0, l), nil
		}
	}

	if m := ctimeAndAsctimeReg.FindStringSubmatch(timeStr); len(m) > 0 {
		var l *time.Location
		if m[6] != "" {
			l = loadLoc(m[6])
		}
		if l == nil {
			l = loc
		}
		return time.Date(adjustYear(m[7]), shortMon2Mon[m[1]], a2i(m[2]), a2i(m[3]), a2i(m[4]), a2i(m[5]), 0, l), nil
	}

	if m := unixLsReg.FindStringSubmatch(timeStr); len(m) > 0 {
		y := a2i(m[3])
		if m[3] == "" {
			y = time.Now().Year()
		}
		return time.Date(y, shortMon2Mon[m[1]], a2i(m[2]), a2i(m[4]), a2i(m[5]), a2i(m[6]), 0, loc), nil
	}

	if m := iso8601Reg.FindStringSubmatch(timeStr); len(m) > 0 {
		var l *time.Location
		if strings.ToLower(m[8]) == "z" {
			l = time.UTC
		} else if m[8] != "" {
			l = time.FixedZone("", offsetStr2offset(m[8]))
		} else {
			l = loc
		}
		nsec := 0
		fracStr := m[7]
		if fracStr != "" {
			of := 9 - len(fracStr)
			if of <= 0 {
				nsec = a2i(fracStr[0:9])
			} else {
				nsec = a2i(fracStr) * int(math.Pow(10, float64(of)))
			}
		}
		return time.Date(a2i(m[1]), time.Month(a2i(m[2])), a2i(m[3]), a2i(m[4]), a2i(m[5]), a2i(m[6]), nsec, l), nil
	}

	if m := winDirReg.FindStringSubmatch(timeStr); len(m) > 0 {
		l := loc
		hr := a2i(m[4])
		ampm := strings.ToUpper(m[6])
		if ampm == "AM" && hr == 12 {
			hr = 0
		} else if ampm == "PM" && hr != 12 {
			hr += 12
		}
		return time.Date(adjustYear(m[3]), time.Month(a2i(m[1])), a2i(m[2]), hr, a2i(m[5]), 0, 0, l), nil
	}

	return time.Time{}, fmt.Errorf("parsing time %q: parse failed", origTimeStr)
}

// Time2Str converts a `time.Time` to a string.
// Note. This is only an implementation as HTTP:Datet#time2str's porting and
// may not be very practical.
//
// The string returned is in the format preferred for the HTTP protocol.
// This is a fixed length subset of the format defined by RFC 1123,
// represented in Universal Time (GMT).  An example of a time stamp
// in this format is:
//
//    Sun, 06 Nov 1994 08:49:37 GMT
func Time2Str(t time.Time) string {
	return t.In(time.FixedZone("GMT", 0)).Format(time.RFC1123)
}
